### PR TITLE
PEP 553 built-in debug() function (bpo-31353)

### PR DESCRIPTION
**Pull request by warsaw**
https://bugs.python.org/issue31353

### CODE REVIEW

**Comment by warsaw** [comment]
bpo-31353

**Comment by warsaw** [comment]
https://www.python.org/dev/peps/pep-0553/

File: Python/sysmodule.c
Comment (freddrake) [inline Python/sysmodule.c:2057]: Your indentation here is different from what's already present here.  While I personally prefer how you've done it, local inconsistency is ugly.

Comment (warsaw) [reply Python/sysmodule.c:2057]: True, but unfortunately that's the only way to not break 80 columns (or maybe "the least horrible way").

Comment (freddrake) [reply Python/sysmodule.c:2057]: Changing the indentation of the other entries should not be discarded as an option if that allows PEP 7 conformance.

Not ideal, perhaps, but allows local consistency.

Comment (warsaw) [reply Python/sysmodule.c:2057]: I'm not sure I want to touch lines outside those related to this PEP.

Comment (freddrake) [reply Python/sysmodule.c:2057]: In the interest of local consistency, that could be done after PEP code is landed (assuming acceptance, which sounds likely), as a separate commit. Otherwise, some way to conform would be valuable.

**Comment by warsaw** [comment]
PEP 553 is accepted.  I'll land this as soon as I get an approval on the PR.

File: Doc/library/sys.rst
Comment (vstinner) [inline Doc/library/sys.rst:135]: PYTHONBREAKPOINT should be written:

```
:envvar:`PYTHONBREAKPOINT`
```

(as you did in the What's New in Python 3.7) and you should document the variable in Doc/using/cmdline.rst.

File: Lib/test/test_builtin.py
Comment (vstinner) [inline Lib/test/test_builtin.py]: Hum, would it be possible to modify the unit test so it doesn't rely on breakpoint() calling pdb.set_trace()? The test may be run with PYTHONBREAKPOINT set to something else.. to debug the test (or another test while running the test suite). Or skip the test if sys.breakpointhook is no sys.__breakpointhook__?

File: Lib/test/test_builtin.py
Comment (vstinner) [inline Lib/test/test_builtin.py]: You might use a mock to get arguments: assert_called_once_with(...).

File: Lib/test/test_builtin.py
Comment (vstinner) [inline Lib/test/test_builtin.py]: I suggest to use support.swap_attr(sys, "breakpointhook", ...): it's shorter and restore the variable to its previous value. Again, the test runner might have replaced sys.breakpointhook for good reasons. Don't reset it to sys.__breakpointhook__.

File: Lib/test/test_builtin.py
Comment (vstinner) [inline Lib/test/test_builtin.py]: Would it be possible to restore sys.breakpointhook to its original value at the end of the test?

File: Lib/test/test_builtin.py
Comment (vstinner) [inline Lib/test/test_builtin.py:1521]: Maybe you should add a setUp() method using self.addCleanup() to save/restore sys.breakpointhook value, instead of doing that in each test.

File: Lib/test/test_builtin.py
Comment (vstinner) [inline Lib/test/test_builtin.py]: env var tests expect that sys.breakpointhook is the default hook. Maybe put these tests in a new test case and set sys.breakpointhook to sys.__breakpointhook__ in setUp() (and use addCleanup to restore the old value)?

Or maybe rewrite the tests to call directly sys.__breakpointhook__ rather than testing indirectly that breakpoint() calls sys.breakpointhook and expect sys.breakpointhook to be sys.__breakpointhook__?

File: Lib/test/test_builtin.py
Comment (vstinner) [inline Lib/test/test_builtin.py]: Would it be possible to add a second V, I'm confused by "envar" for "env var" :-)

File: Misc/NEWS.d/next/Core and Builtins/2017-09-05-14-19-02.bpo-31353.oGZUeJ.rst
Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2017-09-05-14-19-02.bpo-31353.oGZUeJ.rst]: PEP 553 can be written:

```
:pep:`553`
```

to get a direct link to the PEP ;-)

File: Python/bltinmodule.c
Comment (vstinner) [inline Python/bltinmodule.c]: You might try to use the new METH_FASTCALL calling convention.

```
builtin_breakpoint(PyObject *self, PyObject **args, Py_ssize_t nargs, PyObject *kwnames) 
...
PyObject *retval = _PyObject_FastCallKeywords(hook, args, nargs, kwnames);
...
```

It prevents the creation a temporary tuple for 'args' (and dict for 'keywords') and so make your function faster :-D

File: Python/bltinmodule.c
Comment (vstinner) [inline Python/bltinmodule.c]: Why not simply "Call sys.breakpointhook(*args, **kwds)."? It's just a suggestion. I don't know what is the best.

File: Python/sysmodule.c
Comment (vstinner) [inline Python/sysmodule.c]: Hum, you should use Py_GETENV() since PYTHON* env vars should be ignored when using python3 -E.

File: Python/sysmodule.c
Comment (vstinner) [inline Python/sysmodule.c]: Again, you might use METH_FASTCALL.

File: Python/sysmodule.c
Comment (vstinner) [inline Python/sysmodule.c]: Do you want to allow users to modify PYTHONBREAKPOINT env var at runtime? Why not checking PYTHONBREAKPOINT once at startup, somewhere in Py_Initialize()?

File: Python/sysmodule.c
Comment (vstinner) [inline Python/sysmodule.c:125]: *If* you want to handle non-ASCII characters correctly, it seems like env var should be decoded using:

```
#ifdef __APPLE__
            /* Use utf-8 on Mac OS X */
            unicode = PyUnicode_FromString(p);
#else
            unicode = PyUnicode_DecodeLocale(p, "surrogateescape");
#endif
```

according to Modules/main.c code to handle PYTHONWARNINGS. In fact, it's even worse, this file also uses the following code on Windows to get directly Unicode:

```
!Py_IgnoreEnvironmentFlag && (wp = _wgetenv(L"PYTHONWARNINGS"))
```

Maybe leave it unchanged. It's ok to not support non-ASCII Python module/function names on non-UTF-8 locale encodings (really weird case).

File: Python/sysmodule.c
Comment (vstinner) [inline Python/sysmodule.c]: Add "assert(!PyErr_Occurred());" at the entry point since this function is called when something goes wrong, and the implementation don't expect that an exception is currently set.

... Maybe even raise an exception at runtime if it's called with an exception set? See _Py_CheckFunctionResult() which implements a similar check at *exit* point.

File: Python/sysmodule.c
Comment (vstinner) [inline Python/sysmodule.c:159]: I suggest to not dump the variable value to avoid encoding/escaping issues (%s doesn't escape quotes). It's up to you.

File: Python/sysmodule.c
Comment (vstinner) [inline Python/sysmodule.c]: I would prefer to handle the error first "if (status < 0) { return NULL; }", and use "Py_RETURN_NONE" for the regular case. (Except true and false branches.)

File: Python/sysmodule.c
Comment (vstinner) [inline Python/sysmodule.c]: It's unclear to me if the prototype is breakpointhook() or breakpointhook(*args, **kw)?

File: Python/sysmodule.c
Comment (vstinner) [inline Python/sysmodule.c]: I suggest to repeat the first sentence of its doc:

```
This hook function is called by built-in breakpoint().
```

**Review by vstinner — CHANGES_REQUESTED** [review]
You requested my review. Here you have, enjoy! :-)

My main question is if PYTHONBREAKPOINT should be evaluated at each sys.__breakpoint__() call, or if it could be read only once at Python startup?

The PEP explicitly says "PYTHONBREAKPOINT is re-interpreted every time sys.breakpointhook() is reached.". That's surprising to me, but I don't really care and the PEP is already accepted. So please ignore my comments related to this ;-)

File: Doc/library/sys.rst
Comment (warsaw) [reply Doc/library/sys.rst:135]: Thanks, fixed!

File: Lib/test/test_builtin.py
Comment (warsaw) [reply Lib/test/test_builtin.py:1521]: Yes, this is a great suggestion, coupled with the observation below that the test suite could be run with `PYTHONBREAKPOINT` or `sys.breakpointhook` set, we really need the `setUp()` to clean the slate for the subsequent tests.

I'm going to use an `ExitStack()` to do this in the `setUp()` with a `addCleanup()` to close the context manager and restore the original state.

File: Lib/test/test_builtin.py
Comment (warsaw) [reply Lib/test/test_builtin.py]: See above about the clean slate, but in this case we're testing the default behavior, which is exactly that the clean slate calls `pdb.set_trace()`.

File: Lib/test/test_builtin.py
Comment (warsaw) [reply Lib/test/test_builtin.py]: I'm not quite sure what you're suggesting here.  There are other tests to ensure that the right positional and keyword arguments are passed through.  This test just ensures that a custom `sys.breakpointhook` gets called at all.  (I like focused unit tests.)

File: Lib/test/test_builtin.py
Comment (warsaw) [reply Lib/test/test_builtin.py]: Excellent suggestion.  I'm going to use this as mentioned above.

File: Lib/test/test_builtin.py
Comment (warsaw) [reply Lib/test/test_builtin.py]: Yep, see above.

File: Lib/test/test_builtin.py
Comment (warsaw) [reply Lib/test/test_builtin.py]: It's a good suggestion.  I think the clean-slate setup will address this.

File: Lib/test/test_builtin.py
Comment (warsaw) [reply Lib/test/test_builtin.py]: Oh, ha!  "envvar" always looks weird to me.  I've always read/written the shortcut for "environment variable" as "envar" so that's just a habit I always have to break when using `:envvar:` for example. :)

Both are reasonable shortcuts, e.g. **env**ironment **var**iable vs. **en**vironment **var**iable.

If you don't mind, I'll keep it this way.

File: Python/bltinmodule.c
Comment (warsaw) [reply Python/bltinmodule.c]: Heh.  Since this [isn't even documented](https://bugs.python.org/issue28805) I didn't know about them!  I'll think about it (I hate to perpetuate the use of undocumented APIs).

File: Python/sysmodule.c
Comment (warsaw) [reply Python/sysmodule.c]: Yes, we do want to allow that.  Guido even pointed out that use case in one of his messages, and I agree.

File: Python/sysmodule.c
Comment (warsaw) [reply Python/sysmodule.c]: Is that true, i.e. that it's possible to call this with an exception set?  I don't mind adding an assertion for that, but I'm not sure I see how it's possible.  `builtin_breakpoint()` won't call it with an exception set AFAICT.  What scenario are you thinking about?

Comment (vstinner) [reply Python/sysmodule.c]: > Is that true, i.e. that it's possible to call this with an exception set?

In my experience, it's very common that a function returning a result *and* set an exception is not catched immediately, but somewhere "later". I had such issues when I ran my pyfailmalloc tool which injects MemoryError anywhere.

But I made many changes in CPython internals to add assert(!PyErr_Occurred()) at many places. The main change is that all functions calling functions should now check that no exception is raised if the called function returned a result. This is _Py_CheckFunctionResult().

To come back to your breakpoint() implementation, "assert(!PyErr_Occurred());" should be enough.

File: Lib/test/test_builtin.py
Comment (vstinner) [reply Lib/test/test_builtin.py]: I don't mind.

File: Python/sysmodule.c
Comment (warsaw) [reply Python/sysmodule.c]: Just to be clear, you're suggesting `assert(!PyErr_Occurred());` at the beginning of `sys_breakpointhook()` right?  I'm cool with that.

File: Python/sysmodule.c
Comment (warsaw) [reply Python/sysmodule.c]: I wonder if `PYTHONBREAKPOINT` should be ignored with -E.  I suppose it should, but a use case would be a production system that wants a safe environment but *also* wants to ensure that no stray `breakpoint()` calls stop the process.  That was @ncoghlan 's main use case for `PYTHONBREAKPOINT=0`.

Maybe `-E` should effectively set `PYTHONBREAKPOINT=0`?

File: Lib/test/test_builtin.py
Comment (merwok) [reply Lib/test/test_builtin.py]: I think haypo is suggesting code like:

```python
myhook = mock.MagicMock()
sys.breakpointhook = myhook

breakpoint('hello')

myhook.assert_called_once_with('hello')
```

instead of manual tracking of the call status.

File: Lib/test/test_builtin.py
Comment (merwok) [reply Lib/test/test_builtin.py]: What madman shortens environment to en? :)

File: Misc/NEWS.d/next/Core and Builtins/2017-09-05-14-19-02.bpo-31353.oGZUeJ.rst
Comment (merwok) [reply Misc/NEWS.d/next/Core and Builtins/2017-09-05-14-19-02.bpo-31353.oGZUeJ.rst]: Doesn’t docutils auto-linkify text that looks like `RFC 1234` or `PEP 123`?

File: Python/sysmodule.c
Comment (ncoghlan) [reply Python/sysmodule.c]: My suggestion:

- use `Py_GETENV`, so -E ignores `PYTHONBREAKPOINT`, but still defaults to dropping in to pdb
- have -I (isolated mode) imply `PYTHONBREAKPOINT=0` with no way to override it at runtime

That would probably require a field in the CoreConfig struct to say whether or not to even consider checking PYTHONBREAKPOINT for breakpoints or not: https://github.com/python/cpython/blob/master/Include/pystate.h#L27

(Either way, this should probably go back to python-dev for discussion & resolution, since it isn't clear to me what the right answer actually is)

File: Lib/test/test_builtin.py
Comment (warsaw) [reply Lib/test/test_builtin.py]: @merwok ah, thanks!  I like that.

File: Python/sysmodule.c
Comment (warsaw) [reply Python/sysmodule.c:125]: Agreed!  I'll leave this one alone.

File: Python/sysmodule.c
Comment (warsaw) [reply Python/sysmodule.c:159]: For now I think it's useful to include the value.  Let's see if it's a problem in practice.

**Comment by warsaw** [comment]
@haypo @merwok Thanks for the great comments.  I think all issues are address, save the behavior of `PYTHONBREAKPOINT` under `-E`.  Let's see how the python-dev discussion shakes out.

File: Python/sysmodule.c
Comment (vstinner) [reply Python/sysmodule.c]: Yes.

File: Lib/test/test_builtin.py
Comment (vstinner) [reply Lib/test/test_builtin.py]: Ha ha, I proposed the same change, but @warsaw likes "envar" with a single V. It looks a typo to me, but @warsaw likes it. Honestly, since it's just an unit test, I don't care ;-)

File: Doc/using/cmdline.rst
Comment (vstinner) [inline Doc/using/cmdline.rst:513]: ".. versionadded:: 3.7" is missing here.

**Comment by vstinner** [comment]
There is still a bug somewhere in tests:

```
haypo@selma$ PYTHONBREAKPOINT=0 ./python -E -m test test_builtin
Run tests sequentially
0:00:00 load avg: 1.46 [1/1] test_builtin
> /home/haypo/prog/python/master/Lib/test/test_builtin.py(1581)test_envar_good_path_other()
-> mock.assert_called_once_with()
(Pdb) 
```

**Comment by vstinner** [comment]
I ran "./python -m test -R 3:3 test_builtins" and there is no reference leak.

Note: I had to disable two unrelated unit tests because of https://bugs.python.org/issue31703

**Review by vstinner — COMMENTED** [review]
LGTM, but there is a minor issue in tests (see my previous comment) and a missing "verisonadded" in the doc. Moreover, tests are failing, so I cannot approve the change yet.

File: Doc/using/cmdline.rst
Comment (warsaw) [reply Doc/using/cmdline.rst:513]: Fixed

**Comment by warsaw** [comment]
@haypo Yeah, a bunch of tests have to be skipped when `-E` is given!  I don't think there's any way to run them in that case, since `Py_IgnoreEnvironmentFlag` is not settable or mockable from Python.

**Comment by ncoghlan** [comment]
@warsaw There are some other tests that are run in subprocesses specifically to cope with the main test runner being executed with `-E` set. It's less convenient for coverage collection, but avoids the need to skip the tests when `-E` is set.

**Comment by warsaw** [comment]
> I ran "./python -m test -R 3:3 test_builtins" and there is no reference leak.

I couldn't get that to do anything other than just (seemingly) hang on macOS. :(

```
% ./python.exe -m test -R 3:3 test_builtin
./python.exe -m test -R 3:3 test_builtin
Run tests sequentially
0:00:00 load avg: 2.25 [1/1] test_builtin
beginning 6 repetitions
123456
  C-c C-c^C
Test suite interrupted by signal SIGINT.
1 test omitted:
    test_builtin

Total duration: 49 min 14 sec
Tests result: INTERRUPTED
```

Maybe that's caused by the bug you referenced?

**Comment by warsaw** [comment]
@ncoghlan 

> There are some other tests that are run in subprocesses specifically to cope with the main test runner being executed with -E set. It's less convenient for coverage collection, but avoids the need to skip the tests when -E is set.

I think you're talking about `test_cmd_line.py`'s use of `interpreter_requires_environment()`?  I'm not sure this is a big enough win for the complexity.  I'm fine skipping the tests since they really make no sense under `-E`.

**Comment by warsaw** [comment]
@haypo I think we're good to go.  Do you agree?

**Review by vstinner — APPROVED** [review]

**Comment by vstinner** [comment]
Yeah, LGTM. We can still enhance tests later if needed (run tests which depends on -E in subprocesses).

**Pull request merged**
